File: 0817-linked-list-components/0817-linked-list-components.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] nums) {
        Set<Integer>st = new HashSet<>();
        for(int x:nums) {
            st.add(x);
        }
        int flag = 0, ans = 0;
        while(head != null) {
            if(st.contains(head.val)) {
                flag = 1;
            }
            else {
                ans += flag;
                flag = 0;
            }
            head = head.next;
        }
        ans += flag;
        return ans;
    }
}

File: 0817-linked-list-components/0817-linked-list-components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& nums) {
        int ans = 0;
        unordered_set<int>st(nums.begin(), nums.end());
        while(head) {
            if(st.count(head->val) && (head->next == NULL || st.find(head->next->val) == st.end())) ans++;
            head = head->next;
        }
        return ans;
    }
};